wgsl/index.bs

```diff
@@ -18494,6 +18685,559 @@ All synchronization functions [=shader-creation error|must=] only be invoked in
+    <td style="width:10%">Overload
+    <td class="nowrap">
+      <xmp highlight=wgsl>
+        @must_use fn subgroupInclusiveMul(e : T) -> T
+      </xmp>
+  <tr>
+    <td>Preconditions
+    <td>`T` is [=type/concrete=] [=numeric scalar=] or [=numeric vector=]
+  <tr>
+    <td>Description
+    <td>Inclusive scan operation.<br>
+
+    Returns the product of `e` among all [=subgroups/active=] invocations in
+    the [=subgroup=] whose [=subgroup invocation IDs=] are less than or equal
+    to the current invocation's id.
+
+    Note: equivalent to `subgroupExclusiveMul(x) * x`.
+</table>
+
+### `subgroupOr` ### {#subgroupor-builtin}
+
+<table class='data builtin'>
+  <tr algorithm="subgroupOr">
+    <td style="width:10%">Overload
+    <td class="nowrap">
+      <xmp highlight=wgsl>
+        @must_use fn subgroupOr(e : T) -> T
+      </xmp>
+  <tr>
+    <td>Preconditions
+    <td>`T` is [INTEGRAL]
+  <tr>
+    <td>Description
+    <td>Reduction operation.<br>
+
+    Returns the bitwise or (`|`) of `e` among all [=subgroups/active=]
+    invocations in the [=subgroup=].
+</table>
+
+### `subgroupShuffle` ### {#subgroupshuffle-builtin}
+
+<table class='data builtin'>
+  <tr algorithm="subgroupShuffle">
+    <td style="width:10%">Overload
+    <td class="nowrap">
+      <xmp highlight=wgsl>
+        @must_use fn subgroupShuffle(e : T, id : I) -> T
+      </xmp>
+  <tr>
+    <td>Preconditions
+    <td>`T` is [=type/concrete=] [=numeric scalar=] or [=numeric vector=]<br>
+        `I` is [=u32=] or [=i32=]
+  <tr>
+    <td>Description
+    <td>Returns `e` from the invocation whose [=subgroup invocation ID=] matches `id`.
+
+    If `id` is outside the range [0, 128), then:
+    * It is a [=shader-creation error=] if `id` is a [=const-expression=].
+    * It is a [=pipeline-creation error=] if `id` is an [=override-expression=].
+
+    An [=indeterminate value=] is returned if `id` does not select an
+    [=subgroups/active=] invocation.
+</table>
+
+#### `subgroupShuffleDown` #### {#subgroupshuffledown-builtin}
+
+<table class='data builtin'>
+  <tr algorithm="subgroupShuffleDown">
+    <td style="width:10%">Overload
+    <td class="nowrap">
+      <xmp highlight=wgsl>
+        @must_use fn subgroupShuffle(e : T, delta : u32) -> T
+      </xmp>
+  <tr>
+    <td>Preconditions
+    <td>`T` is [=type/concrete=] [=numeric scalar=] or [=numeric vector=]
+  <tr>
+    <td>Description
+    <td>Returns `e` from the invocation whose [=subgroup invocation ID=]
+    matches `subgroup_invocation_id + delta` for the current invocation.
+
+    If `delta` is outside the range [0, 128), then:
+    * It is a [=shader-creation error=] if `delta` is a [=const-expression=].
+    * It is a [=pipeline-creation error=] if `delta` is an [=override-expression=].
+
+    A [=trigger/subgroup_uniformity=] [=diagnostic=] is [=triggered=] if
+    `delta` is not a [=uniform value=].
+    An [=indeterminate value=] is returned if `subgroup_invocation_id + delta`
```

Comment:
I'm trying to understand how one would use this and what's the benefit compared to a `subgroupShuffle`.  It look like a uniform delta would be useful to do a shift, but then the highest active invocation would get an indeterminate value (what we'd want would be a rotation). Should we instead expose some form of shuffleRotate that is easy to use correctly when in subgroup-uniform control flow?

Reply:
It is likely much cheaper on some implementations than a shuffle. I guess it depends on use case. I don't know all the best use cases for this, but some sort of iterative compaction wouldn't care about indeterminate values.

There is a rotate primitive in MSL and SPIR-V, though support in SPIR-V seems rather sparse ([VK_KHR_shader_subgroup_rotate is ~8.3% supported](https://vulkan.gpuinfo.org/listextensions.php)). It could be polyfilled, but it would likely also be more expensive than a pure rotate on some hardware.